src/persist-client/src/impl/machine.rs

```diff
@@ -0,0 +1,293 @@
+        let res = self
+            .apply_unbatched_cmd(deadline, |_, state| {
+                state.downgrade_since(reader_id, new_since)
+            })
+            .await?;
+        let (seqno, _) = match res {
+            Ok(x) => x,
+            Err(err) => return Ok(Err(err)),
+        };
+        Ok(Ok(seqno))
+    }
+
+    pub async fn expire_writer(
+        &mut self,
+        deadline: Instant,
+        writer_id: &WriterId,
+    ) -> Result<SeqNo, LocationError> {
+        let res = self
+            .apply_unbatched_cmd(deadline, |seqno, state| {
+                state.expire_writer(seqno, writer_id)
+            })
+            .await?;
+        let seqno = match res {
+            Ok((seqno, ())) => seqno,
+            Err(NoOp { seqno }) => seqno,
+        };
+        Ok(seqno)
+    }
+
+    pub async fn expire_reader(
+        &mut self,
+        deadline: Instant,
+        reader_id: &ReaderId,
+    ) -> Result<SeqNo, LocationError> {
+        let res = self
+            .apply_unbatched_cmd(deadline, |seqno, state| {
+                state.expire_reader(seqno, reader_id)
+            })
+            .await?;
+        let seqno = match res {
+            Ok((seqno, ())) => seqno,
+            Err(NoOp { seqno }) => seqno,
+        };
+        Ok(seqno)
+    }
+
+    pub async fn snapshot(
+        &mut self,
+        deadline: Instant,
+        as_of: &Antichain<T>,
+    ) -> Result<Result<Vec<String>, InvalidUsage>, LocationError> {
+        // TODO: This fetches the latest state and uses that to determine if we
+        // can serve `as_of`. It's possible that this `as_of` could be served by
+        // the version we happen to have cached, but not by this one we fetch.
+        // Do we care?
+        self.fetch_and_update_state(deadline).await?;
+        Ok(self.state.snapshot(as_of))
+    }
+
+    pub async fn next_listen_batch(
+        &mut self,
+        deadline: Instant,
+        frontier: &Antichain<T>,
+    ) -> Result<(Vec<String>, Description<T>), LocationError> {
+        loop {
+            self.fetch_and_update_state(deadline).await?;
+            if let Some((keys, desc)) = self.state.next_listen_batch(frontier) {
+                return Ok((keys.to_owned(), desc.clone()));
+            }
+            // Wait a bit and try again.
+            //
+            // TODO: See if we can watch for changes in Consensus to be more
+            // reactive here.
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+    }
+
+    async fn apply_unbatched_cmd<
+        R,
+        E,
+        WorkFn: FnMut(SeqNo, &mut State<K, V, T, D>) -> Result<R, E>,
+    >(
+        &mut self,
+        deadline: Instant,
+        mut work_fn: WorkFn,
+    ) -> Result<Result<(SeqNo, R), E>, LocationError> {
+        loop {
+            let new_seqno = self.seqno.unwrap_or_default().next();
+            let mut new_state = self.state.clone();
+            let work_ret = match work_fn(new_seqno, &mut new_state) {
```

Comment:
just want to double check the mechanics here:

does it ever make sense for `work_fn` to succeed on one iteration of the loop and then fail on retry? I want to say .. only in cases when a reader/writer is expired in between contending for a cas